Docs: plugin migration guide 7.0.0 to 8.0.0 (#38911)
grafana/grafana: docs/sources/developers/plugins/migration-guide.md

# Plugin migration guide

## Introduction

This guide explains how to migrate Grafana plugins from previous version to the latest version available. It is structured in a way where you easily should be able to identify what steps you need to take given the Grafana version your plugin currently support.

> If you've successfully migrated your plugin by using this guide and think something is missing, please [submit an issue on GitHub](https://github.com/grafana/grafana/issues/new?title=Docs%20feedback:%20/developers/plugins/migration-guide.md) and share your experiences with us so that we can improve this guide!

## Table of contents

- [From version 7.x.x to 8.0.0](#from-version-7xx-to-800)
  - [Backend plugin v1 support has been dropped](#backend-plugin-v1-support-has-been-dropped)
    - [1. Add dependency on grafana-plugin-sdk-go](#1-add-dependency-on-grafana-plugin-sdk-go)

>     - [2. Update the way you bootrap your plugin](#2-update-the-way-you-bootrap-your-plugin)

Reviewer: ```suggestion
    - [2. Update the way you bootstrap your plugin](#2-update-the-way-you-bootstrap-your-plugin)
```

Author: haha - but its a boot rap...

![bootrap](https://media.giphy.com/media/26FmQPtkD2oatMC9G/giphy.gif?cid=ecf05e47uj7cnsouncw64ksb5ylngng18vykwdeo0bdopyyc&rid=giphy.gif&ct=g)

  - [What's new in Grafana 7.0?](#whats-new-in-grafana-70)
  - [Migrate a plugin from Angular to React](#migrate-a-plugin-from-angular-to-react)
    - [Migrate a panel plugin](#migrate-a-panel-plugin)
    - [Migrate a data source plugin](#migrate-a-data-source-plugin)
    - [Migrate to data frames](#migrate-to-data-frames)
  - [Troubleshoot plugin migration](#troubleshoot-plugin-migration)

## From version 7.x.x to 8.x.x

This guide will help you migrate Grafana v7.x.x plugins to the updated plugin system released with Grafana v8.x.x. All the changes described below might not be applicable to your plugin but we will try to cover all breaking changes in Grafana v8.x.x and what steps you need to take to upgrade your plugin.

### Backend plugin v1 support has been dropped

To get your backend plugin running in Grafana 8 you need to use the new [plugin sdk](https://github.com/grafana/grafana-plugin-sdk-go).

#### 1. Add dependency on grafana-plugin-sdk-go

Add a dependency on the `https://github.com/grafana/grafana-plugin-sdk-go`. We recommend using [go modules](https://go.dev/blog/using-go-modules) to manage your dependencies.

> #### 2. Update the way you bootrap your plugin

Reviewer: ```suggestion
#### 2. Update the way you bootstrap your plugin
```

| time                 | temperature |
| -------------------- | ----------- |
| 2021-05-08T07:31:45Z | 23          |
| 2021-05-08T09:31:45Z | 25          |
| 2021-05-08T11:31:45Z | 27          |

| time                 | humidity |
| -------------------- | -------- |
| 2021-05-08T07:31:45Z | 60       |
| 2021-05-08T09:31:45Z | 55       |
| 2021-05-08T11:31:45Z | 55       |

> Data delivered as an array of data frames (time field needs to be repeated).

That made it possible to detect time series data by inspecting the data frame. If it had two fields (time + value) it most likley was time series data. This is no longer possible since the same data can be delivered in the following `wide` format.

| time                 | temperature | humidity |
| -------------------- | ----------- | -------- |
| 2021-05-08T07:31:45Z | 23          | 60       |
| 2021-05-08T09:31:45Z | 25          | 55       |
| 2021-05-08T11:31:45Z | 27          | 55       |

> Data delivered as one single data frame (the values share the same time field).

If your plugin should support time series data we recommend you to make sure to support both.

> ### Update react-hook-form from v6 to v7

Author: @leventebalogh I just referenced the `react-hook-form` migration guide. I was thinking that if there are things that is a bit unclear in that guide we might add a section here that highlights how to resolve those.

WDYT?

Reviewer 2: I think it's great bud. 

It was actually painful for us because we were using `react-hook-form` in our plugins directly in a lot of places, and  those needed to be updated. 

You are right, the best is if we are just linking their original migration guide. 👍

> We have upgrade the react-hook-form from version 6 to version 7. We recommend follow the [react-hook-form-migration-guide](#https://react-hook-form.com/migrate-v6-to-v7/) to get make your forms compatible with that version.

Reviewer 2: **Nit:** two small typos.
```suggestion
We have upgraded the react-hook-form from version 6 to version 7. We recommend follow the [react-hook-form-migration-guide](https://react-hook-form.com/migrate-v6-to-v7/) to get make your forms compatible with that version.
```

### Update the plugin.json

> The property of defining what Grafana version your plugin support has been renamed.

Reviewer 2: **Nit:** small typo
```suggestion
The property of defining what Grafana version your plugin supports has been renamed.
```

```json
// before
{
"dependencies": {
    "grafanaVersion": "7.5.x",
    "plugins": []
  }
}

// after
{
  "dependencies": {
    "grafanaDependency": "8.0.0",
    "plugins": []
  }
}
```

### Update imports to match emotion 11

Grafana uses emotion to manage the styling of the frontend. The emotion package has now been updated which might affect your frontend plugin if you have any custom styling in it. Luckily you only need to update the import statements to get it working in Grafana 8.

```ts
// before
import { cx, css } from 'emotion';

// after
import { cx, css } from '@emotion/css';
```

### 8.0 deprecations

> #### Grafana theme v1

Author: @jackw do you think we should cover more things in this section?

Reviewer: @mckn the only other thing I can think of that we might consider adding here is the addition of `useStyles2`? 🤔

Author: ah, that is true!

> In Grafana 8 we have introduced a new improved version of our theming system. The previous version of the theming system is still available but deprecated and will be removed in the next major version of Grafana.

Reviewer 2: **Nit:** maybe sounds better like this?
```suggestion
In Grafana 8 we have introduced a new improved version of our theming system. The previous version of the theming system is still available but is deprecated and will be removed in the next major version of Grafana.
```

```ts
// before
import React, { ReactElement } from 'react';
import { useTheme } from '@grafana/ui';

function Component(): ReactElement | null {
  const theme = useTheme();
}

// after
import React, { ReactElement } from 'react';
import { useTheme2 } from '@grafana/ui';

function Component(): ReactElement | null {
  const theme = useTheme2();
  // Your component has access to the theme variables now
}
```

**How to use the theme in a class component**

```ts
// before
import React from 'react';
import { Themeable, withTheme } from '@grafana/ui';

type Props = {} & Themeable;

class Component extends React.Component<Props> {
  render() {
    const { theme } = this.props;
    // Your component has access to the theme variables now
  }
}

export default withTheme(Component);

// after
import React from 'react';
import { Themeable2, withTheme2 } from '@grafana/ui';

type Props = {} & Themeable2;

class Component extends React.Component<Props> {
  render() {
    const { theme } = this.props;
    // Your component has access to the theme variables now
  }
}

export default withTheme2(Component);
```

**Gradually migrating components**

> If you end up in a situation where you need to use both the v1 and v2 theme due to using migraded and non-migrated components in the same context you can handle that by using the `v1` property on the `v2` theme as described below.

Reviewer 2: **Nit:** small typo
```suggestion
If you end up in a situation where you need to use both the v1 and v2 theme due to using migrated and non-migrated components in the same context you can handle that by using the `v1` property on the `v2` theme as described below.
```

	return &SampleDatasource{}, nil
}

type SampleDatasource struct{}

func (d *SampleDatasource) Dispose() {
	// Clean up datasource instance resources.
}

func (d *SampleDatasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
  log.DefaultLogger.Info("QueryData called", "request", req)
  // logic for querying your datasource.
}

func (d *SampleDatasource) CheckHealth(_ context.Context, req *backend.CheckHealthRequest) (*backend.CheckHealthResult, error) {
  log.DefaultLogger.Info("CheckHealth called", "request", req)
  // The main use case for these health checks is the test button on the
  // datasource configuration page which allows users to verify that
  // a datasource is working as expected.
}
```

### Unsigned backend plugins will not be loaded

Before you can run a backend plugin on Grafana installations, you must sign the plugin. Grafana Labs strongly recommends you upload signed plugins since we cannot guarantee the integrity of unsigned plugins.

For instructions on plugin signing, refer to the [Plugin signatures](https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/#sign-a-plugin) topic.

> You can still run and develop an unsigned plugin by running your Grafana instance in [development mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode).

Reviewer 3: I don't think we want to encourage anyone to run their Grafana instance in anything other than `production`. I think using [this](https://grafana.com/docs/grafana/latest/administration/configuration/#allow_loading_unsigned_plugins) config is the most appropriate way to go for now

Reviewer 3: ```suggestion
You can still run and develop an unsigned plugin by running your Grafana instance in [development mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode). Alternatively, you can use [this configuration setting.]({{< relref "../administration/#allow_loading_unsigned_plugins" >}})
```

Reviewer 2: Yea, I think that is a good addition, thanks @wbrowne!

> You can still run and develop an unsigned plugin by running your Grafana instance in [development mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode). Alternatively, you can use [this configuration setting.]({{< relref "../administration/#allow_loading_unsigned_plugins" >}})

Reviewer 4: ```suggestion
You can still run and develop an unsigned plugin by running your Grafana instance in [development mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode). Alternatively, you can use the [allow_loading_unsigned_plugins configuration setting.]({{< relref "../administration/#allow_loading_unsigned_plugins" >}})
```

### Update react-hook-form from v6 to v7

> We have upgraded react-hook-form from version 6 to version 7. We recommend following the [react-hook-form-migration-guide](https://react-hook-form.com/migrate-v6-to-v7/) to make your forms compatible with version 7.

Reviewer 4: ```suggestion
We have upgraded react-hook-form from version 6 to version 7. To make your forms compatible with version 7, refer to the [react-hook-form-migration-guide](https://react-hook-form.com/migrate-v6-to-v7/).
```

```ts
// before
import React, { ReactElement } from 'react';
import { useTheme } from '@grafana/ui';

function Component(): ReactElement | null {
  const theme = useTheme();
}

// after
import React, { ReactElement } from 'react';
import { useTheme2 } from '@grafana/ui';

function Component(): ReactElement | null {
  const theme = useTheme2();
  // Your component has access to the theme variables now
}
```

**How to use the theme in a class component**

```ts
// before
import React from 'react';
import { Themeable, withTheme } from '@grafana/ui';

type Props = {} & Themeable;

class Component extends React.Component<Props> {
  render() {
    const { theme } = this.props;
    // Your component has access to the theme variables now
  }
}

export default withTheme(Component);

// after
import React from 'react';
import { Themeable2, withTheme2 } from '@grafana/ui';

type Props = {} & Themeable2;

class Component extends React.Component<Props> {
  render() {
    const { theme } = this.props;
    // Your component has access to the theme variables now
  }
}

export default withTheme2(Component);
```

**Gradually migrating components**

> If you end up in a situation where you need to use both the v1 and v2 theme due to using migrated and non-migrated components in the same context you can handle that by using the `v1` property on the `v2` theme as described below.

Reviewer 4: ```suggestion
If you need to use both the v1 and v2 themes due to using migrated and non-migrated components in the same context, use the `v1` property on the `v2` theme as described in the following example.
```

# Plugin migration guide

## Introduction

> This guide explains how to migrate Grafana plugins from previous version to the latest version available. It is structured in a way where you easily should be able to identify what steps you need to take given the Grafana version your plugin currently support.

Reviewer 4: Can we define "latest", otherwise it will be stale in a future release,

    - [3. Update the plugin package](#3-update-the-plugin-package)
  - [Unsigned backend plugins will not be loaded](#unsigned-backend-plugins-will-not-be-loaded)
  - [Time series data can now be in wide or many format](#time-series-data-can-now-be-in-wide-or-many-format)
  - [Update react-hook-form from v6 to v7](#update-react-hook-form-from-v6-to-v7)
  - [Update the plugin.json](#update-the-pluginjson)
  - [Update imports to match emotion 11](#update-imports-to-match-emotion-11)
  - [8.0 Deprecations](#80-deprecations)
    - [Grafana theme v1](#grafana-theme-v1)
- [From version 6.2.x to 7.4.0](#from-version-62x-to-740)
  - [Legend components](#legend-components)
- [From version 6.x.x to 7.0.0](#from-version-6xx-to-700)
  - [What's new in Grafana 7.0?](#whats-new-in-grafana-70)
  - [Migrate a plugin from Angular to React](#migrate-a-plugin-from-angular-to-react)
    - [Migrate a panel plugin](#migrate-a-panel-plugin)
    - [Migrate a data source plugin](#migrate-a-data-source-plugin)
    - [Migrate to data frames](#migrate-to-data-frames)
  - [Troubleshoot plugin migration](#troubleshoot-plugin-migration)

## From version 7.x.x to 8.x.x

> This guide will help you migrate Grafana v7.x.x plugins to the updated plugin system released with Grafana v8.x.x. All the changes described below might not be applicable to your plugin but we will try to cover all breaking changes in Grafana v8.x.x and what steps you need to take to upgrade your plugin.

Reviewer 4: ```suggestion
This section explains how to migrate Grafana v7.x.x plugins to the updated plugin system available in Grafana v8.x.x. Depending on your plugin, you need to perform  one or more of the following steps. We have documented the breaking changes in Grafana v8.x.x and the steps you need to take to upgrade your plugin.
```

### Backend plugin v1 support has been dropped

> To get your backend plugin running in Grafana 8 you need to use the new [plugin sdk](https://github.com/grafana/grafana-plugin-sdk-go).

Reviewer 4: ```suggestion
Use the new [plugin sdk](https://github.com/grafana/grafana-plugin-sdk-go) to run your backend plugin running in Grafana 8.
```

Reviewer 5: Change sdk to SDK.

#### 1. Add dependency on grafana-plugin-sdk-go

Add a dependency on the `https://github.com/grafana/grafana-plugin-sdk-go`. We recommend using [go modules](https://go.dev/blog/using-go-modules) to manage your dependencies.

> #### 2. Update the way you bootstrap your plugin

Reviewer 6: I love this in-depth explainer, with code samples (!!!), of how to migrate. Perhaps it could be better suited to a subpage (`/migration-guide/updating-to-plugin-sdk` or something), with a brief summary of the change, and why.

Reviewer 2: Thanks @joshhunt!

Regarding the why - I think the main reason behind the plugin SDK was to reduce the boilerplate and make it easier to just spin up a backend plugin. Will check a bit more though on what our direction is with the SDK.

The subpage is not a bad idea, we are also thinking on how we could better organise our migration guides (and also the other plugin related docs). We probably won't do it in this PR, but will let you and the other frontenders as soon as we have a design doc.

  // logic for querying your datasource.
}

// after
package plugin

import (
  "context"

  "github.com/grafana/grafana-plugin-sdk-go/backend"
  "github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
  "github.com/grafana/grafana-plugin-sdk-go/backend/log"
  "github.com/grafana/grafana-plugin-sdk-go/data"
)

func NewSampleDatasource(_ backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
	return &SampleDatasource{}, nil
}

type SampleDatasource struct{}

func (d *SampleDatasource) Dispose() {
	// Clean up datasource instance resources.
}

func (d *SampleDatasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
  log.DefaultLogger.Info("QueryData called", "request", req)
  // logic for querying your datasource.
}

func (d *SampleDatasource) CheckHealth(_ context.Context, req *backend.CheckHealthRequest) (*backend.CheckHealthResult, error) {
  log.DefaultLogger.Info("CheckHealth called", "request", req)
  // The main use case for these health checks is the test button on the
  // datasource configuration page which allows users to verify that
  // a datasource is working as expected.
}
```

> ### Unsigned backend plugins will not be loaded

Author: hmm, would `Backend plugins loading` work? That is basically the area we are covering and adding that the plugin needs to be signed to be able to run on all Grafana instances.

> ### Unsigned backend plugins will not be loaded

Reviewer 2: I would lean towards the original version as well in this case (`Unsigned backend plugins will not be loaded`), I think it is giving a good excerpt on what is changing here.

> ### Unsigned backend plugins will not be loaded

Reviewer 5: ```suggestion
### Sign and load backend plugins
```
See my subsequent suggestion.

> ### Unsigned backend plugins will not be loaded

Reviewer 5: See my recommendations at https://github.com/grafana/grafana/pull/38911#discussion_r709039652 
and https://github.com/grafana/grafana/pull/38911#discussion_r709035546.

> We strongly recommend our Grafana users not to allow unsigned plugins in their Grafana installation. By allowing unsigned plugins, we can’t guarantee the authenticity of the plugin which could compromise the security of your Grafana installation.

Reviewer 4: ```suggestion
Before you can run your plugin on Grafana installations, you must sign the plugin. Grafana Labs strongly recommends you upload signed plugins. We cannot guarantee the integrity of unsigned plugins.
```

> We strongly recommend our Grafana users not to allow unsigned plugins in their Grafana installation. By allowing unsigned plugins, we can’t guarantee the authenticity of the plugin which could compromise the security of your Grafana installation.

Reviewer 5: ```suggestion
Unsigned backend plugins will not be loaded.

We strongly recommend that you not allow unsigned plugins in your Grafana installation. By allowing unsigned plugins, we cannot guarantee the authenticity of the plugin, which could compromise the security of your Grafana installation.
```
This indirect language is odd, unless we are writing to a third-party in a conversation. Is the user the one reading this or is something else reading and relaying the information to the user?

Reviewer 2: Yea, I think the user is the one reading this 🤔

Reviewer 2: Aaa it has already been updated based on @achatterjee-grafana's comments.

Reviewer 2: Applied your suggestion in https://github.com/grafana/grafana/pull/38911/commits/e11fa23f25d26ce162ddb4432fe2c6d909f01176.

> This means that you, as a plugin developer, need to get your plugin signed if you want it to be able to run on all Grafana installations.

Reviewer 4: Remove.

> This means that you, as a plugin developer, need to get your plugin signed if you want it to be able to run on all Grafana installations.
> 
> Follow the following [steps](https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/#sign-a-plugin) to get instructions on how to set up plugin signing.

Reviewer 4: ```suggestion
For instructions on plugin signing, refer to the [Plugin signatures](https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/#sign-a-plugin) topic.
```

> Follow the following [steps](https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/#sign-a-plugin) to get instructions on how to set up plugin signing.

Reviewer 5: ```suggestion
To sign your plugin, see [Sign a plugin](https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/#sign-a-plugin).
```

Reviewer 2: Thanks @osg-grafana, although it seems that this was already changed based on @achatterjee-grafana's suggestion before: https://github.com/grafana/grafana/pull/38911#discussion_r705557323

Reviewer 5: One of us can always go in later and do a second review. All good. "Follow the following ..." seems bloated.

Reviewer 2: Applied your suggestion in https://github.com/grafana/grafana/pull/38911/commits/e11fa23f25d26ce162ddb4432fe2c6d909f01176.

### Time series data can now be structured in a wide or many format.

> In Grafana 8.x, you can structure time series data in either `wide` or `many` formats. If your plugin supports time series data, we recommend that you support both formats.

Reviewer 6: Is this for visualisations, or data source plugins?

Author: panel plugins

Reviewer 2: Resolving this one as this part got removed from the guide. ([More info and explanation](https://github.com/grafana/grafana/pull/38911#discussion_r708992021))

> You will still be able to run and develop your plugin during development by running your Grafana [instance in development mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode).

Reviewer 4: ```suggestion
After you sign, you can still run and develop your plugin by running your Grafana instance in [development mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode).
```

Reviewer: I'm not sure this suggested change is correct as it implies you need to sign the plugin at least once before being at to run it in development mode. However to the best of my knowledge an unsigned plugin can be run in development mode without it ever being signed.

Reviewer 4: How about:
You can still run and develop an unsigned plugin by running your Grafana instance in [development mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode).
```

> ### Time series data can now be structured in a wide or many format.

Author: Would something like this work?

`Support time series data in both `wide` and `many` format"

We basically want people to start support both formats in their plugins. Grafana will provide the panel plugins with time series data in either the wide- or many format. So the plugin developers need to make sure they can handle both types.

> ### Time series data can now be structured in a wide or many format.

Reviewer 2: Resolving this one as this part got removed from the guide. ([More info and explanation](https://github.com/grafana/grafana/pull/38911#discussion_r708992021))

Grafana supported only the `many` format before. For example:

| time                 | temperature |
| -------------------- | ----------- |
| 2021-05-08T07:31:45Z | 23          |
| 2021-05-08T09:31:45Z | 25          |
| 2021-05-08T11:31:45Z | 27          |

| time                 | humidity |
| -------------------- | -------- |
| 2021-05-08T07:31:45Z | 60       |
| 2021-05-08T09:31:45Z | 55       |
| 2021-05-08T11:31:45Z | 55       |

> > Data is delivered as an array of data frames where the `time` field is repeated.

Reviewer 6: I'm not sure if it's entirely clear here that these two tables represent multiple frames, with two fields each, in the same response (??). Though I'm not sure how else to structure it (i find data frames to be extremely difficult to visualise).

Perhaps instead it could be visualised as code? Dunno.

Author: Good point! Should I try to add it as `json` instead?

Reviewer 6: Maybe as like Go or Javascript code that actually constructs the frames? I'm not sure if that's more obvious and easier to read? Just  throwing an idea out there.

Reviewer 2: Resolving this one as this part got removed from the guide. ([More info and explanation](https://github.com/grafana/grafana/pull/38911#discussion_r708992021))

> That made it possible to detect time series data by inspecting the data frame. If it had two fields (time + value) it most likley was time series data. This is no longer possible since the same data can be delivered in the following `wide` format.

Reviewer 4: Can you clarify what the new detection method should be? This will help the user understand what to do instead.

> That made it possible to detect time series data by inspecting the data frame. If it had two fields (time + value) it most likley was time series data. This is no longer possible since the same data can be delivered in the following `wide` format.

Author: So there are basically no new detection method. They need to support both formats.

> That made it possible to detect time series data by inspecting the data frame. If it had two fields (time + value) it most likley was time series data. This is no longer possible since the same data can be delivered in the following `wide` format.

Reviewer 6: I'm confused by the last sentence. Is it saying that _Grafana_ can no longer detect time series data because it has more support for wide format?

I like the context provided by this, explaining _why_ things are the way they are, but I would reword this to focus very explicitly on _Grafana_ behaviour (old and new) and expectations of the plugin developer.

> That made it possible to detect time series data by inspecting the data frame. If it had two fields (time + value) it most likley was time series data. This is no longer possible since the same data can be delivered in the following `wide` format.

Reviewer 2: I think the main reason behind moving in this direction was that with the `wide` time-series representation multiple data frames can be "merged" in case the values have the same timestamps, and thus the size of the representation will be smaller.

While using the `many` format (one value per timestamp) it was easy to say that the data frame represents a time-series in case there were two values in a row: a timestamp and a value. However with the introduction of the `wide` representation (multiple metrics for a single timestamp) this heuristic doesn't work anymore.

> That made it possible to detect time series data by inspecting the data frame. If it had two fields (time + value) it most likley was time series data. This is no longer possible since the same data can be delivered in the following `wide` format.

Reviewer 2: Resolving this one as this part got removed from the guide. ([More info and explanation](https://github.com/grafana/grafana/pull/38911#discussion_r708992021))

| time                 | temperature | humidity |
| -------------------- | ----------- | -------- |
| 2021-05-08T07:31:45Z | 23          | 60       |
| 2021-05-08T09:31:45Z | 25          | 55       |
| 2021-05-08T11:31:45Z | 27          | 55       |

> Data delivered as one single data frame (the values share the same time field).

If your plugin should support time series data we recommend you to make sure to support both.

> 

Reviewer 6: Why would I support multiple time series formats? What's the point of that?

_How_ do I support multiple formats? Can I just return all formats from my data source and be done with it?

Author: That is basically target to panel plugins. The reason why we need to suport both `wide` and `many` is because depending on the data source you will receive time series data in one of the formats. So if you want to support all different kinds of data sources we recommend you to handle time series data in both the wide and many format.

Reviewer 6: I think that entire comment could be in these docs to provide context

 - this is for panel plugins
 - why: because data sources can return both formats, and for your panel to support all data sources it should support both wide and many.

Reviewer 2: Good suggestions! 

After discussing this with the team shortly **I think we are removing the time-series part from the migration guide for now.**

The reason being is that we need a bit more time to understand motivations behind these changes, and then put together something that clearly explains who should be concerned about this change, and what are the actual steps that need to be followed in certain scenarios.

Thanks a lot for the feedback, we will link in the comments in the upcoming separate PR.

> If your plugin should support time series data we recommend you to make sure to support both.

Reviewer 2: Resolving this one as this part got removed from the guide. ([More info and explanation](https://github.com/grafana/grafana/pull/38911#discussion_r708992021))

### Update react-hook-form from v6 to v7

> We have upgraded the react-hook-form from version 6 to version 7. We recommend follow the [react-hook-form-migration-guide](https://react-hook-form.com/migrate-v6-to-v7/) to get make your forms compatible with that version.

Reviewer: ```suggestion
We have upgraded react-hook-form from version 6 to version 7. We recommend following the [react-hook-form-migration-guide](https://react-hook-form.com/migrate-v6-to-v7/) to make your forms compatible with version 7.
```

### Update the plugin.json

The property that defines which Grafana version your plugin supports has been renamed.

```json
// before
{
"dependencies": {
    "grafanaVersion": "7.5.x",
    "plugins": []
  }
}

// after
{
  "dependencies": {

>     "grafanaDependency": "8.0.0",

Reviewer 7: The format should be `grafanaDependency: '>=8.0.0`

Also,  It seems the plugin website still uses the `grafanaVersion` as shown in the [servicenow datasource documentation](https://grafana.com/grafana/plugins/grafana-servicenow-datasource/) whereas the saphana plugin.json uses `grafanaDependency` (new property) but the [plugin website](https://grafana.com/grafana/plugins/grafana-saphana-datasource/) doesn't honor that. It seems we need to fix the website to show the correct dependency version from `grafanaDependency`.

Author: Yes, great feedback! Will look into this 🙏🏻

Reviewer 2: Exactly, it seems like the website template is only looking for the `grafanaVersion` property at the moment, which is not correct. ([Source code](https://github.com/grafana/website/blob/7fb3e6dbdffef62f1f0f5582581b4492091ccc66/layouts/grafana/plugin.html#L100))

I think it should be taking the `grafanaDependency` and falling back to `grafanaVersion` for backward compatibility.
Will create an issue for this.

**Also updated the code sample in the migration guide to use a range.**
